### PR TITLE
Custom rest_client_options via Conjur configuration

### PR DESCRIPTION
**Pull request by doodlesbykumbi**
### What does this PR do?

This replaces the monkey patch to RestClient::Request that made it always uses :ssl_cert_store, which polluted into usage of RestClient outside the scope of Conjur. The new approach provides a sane default that retains the previous behavior, and gives the user the ability to configure RestClient::Requests scoped only to Conjur.

### What ticket does this PR close?
Connected to https://github.com/cyberark/conjur-api-ruby/issues/188

#### Change log
- [x] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [x] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [x] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [ ] This PR does not require updating any documentation

### CODE REVIEW

**Comment by fmang** [comment]
That does the trick for me. I checked that the previous approach with `OpenSSL::SSL::SSLContext::DEFAULT_CERT_STORE.add_file` still works, and that the new approach with `Conjur.configuration.rest_client_options = { ssl_ca_file: 'my_cert.crt' }` works too. I also checked with a proxy.

Thank you !

**Review by diverdane — COMMENTED** [review]
Certainly LGTM, but could probably use a review by someone more familiar with this code.

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,18 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
 
+### Added
+- Add `rest_client_options` option to `Conjur.configuration`
+
+### Changed
+- Replaces monkey patching RestClient::Request, to always set :ssl_cert_store,
```
Comment (jtuttle) [inline CHANGELOG.md:13]: Can we shorten this CHANGELOG entry a bit? Something like "Remove RestClient::Request monkeypatch to allow configuration of :ssl_cert_store to be scoped to Conjur".

File: README.md [hunk 2]
```diff
@@ -107,6 +107,10 @@ For example, specify the `account` and `appliance_url` (both of which are requir
 ```
 Conjur.configuration.account = 'my-account'
 Conjur.configuration.appliance_url = 'https://conjur.mydomain.com/api'
+Conjur.configuration.rest_client_options = {
```
Comment (jtuttle) [inline README.md:110]: Can you add a blurb about why we have to do this? Is it required?

File: lib/conjur/api/router/v4.rb [hunk 3]
```diff
@@ -8,18 +8,18 @@ module V4
 
         def authn_login account, username, password
           verify_account(account)
-          RestClient::Resource.new(Conjur.configuration.authn_url, user: username, password: password)['users/login']
+          RestClient::Resource.new(Conjur.configuration.authn_url, Conjur.configuration.rest_client_options.merge(user: username, password: password))['users/login']
```
Comment (jtuttle) [inline lib/conjur/api/router/v4.rb:11]: Please break this line up and maybe add a helper variable for `Conjur.configuration.rest_client_options.merge(...)`

File: lib/conjur/api/router/v4.rb [hunk 4]
```diff
@@ -8,18 +8,18 @@ module V4
...
         end
 
         def authn_authenticate account, username
           verify_account(account)
-          RestClient::Resource.new(Conjur.configuration.authn_url)['users'][fully_escape username]['authenticate']
+          RestClient::Resource.new(Conjur.configuration.authn_url, Conjur.configuration.rest_client_options)['users'][fully_escape username]['authenticate']
```
Comment (jtuttle) [inline lib/conjur/api/router/v4.rb:16]: Again please break this line up.

File: lib/conjur/api/router/v4.rb [hunk 5]
```diff
@@ -33,31 +33,31 @@ def authn_rotate_api_key credentials, account, id
           else
             [ id.kind, id.identifier ].join('/')
           end
-          RestClient::Resource.new(Conjur.configuration.authn_url, credentials)['users']["api_key?id=#{username}"]
+          RestClient::Resource.new(Conjur.configuration.authn_url, Conjur.configuration.rest_client_options.merge(credentials || {}))['users']["api_key?id=#{username}"]
```
[the remark this answers is not in the document]
Comment (jtuttle) [reply lib/conjur/api/router/v4.rb:36]: Please break this line up too.

File: lib/conjur/api/router/v4.rb [hunk 6]
```diff
@@ -33,31 +33,31 @@ def authn_rotate_api_key credentials, account, id
...
           http_options = {
             headers: { authorization: %Q(Token token="#{token}") }
           }
-          RestClient::Resource.new(Conjur.configuration.core_url, http_options)['host_factories']['hosts']
+          RestClient::Resource.new(Conjur.configuration.core_url, Conjur.configuration.rest_client_options.merge(http_options || {}))['host_factories']['hosts']
         end
 
         def host_factory_create_tokens credentials, id
-          RestClient::Resource.new(Conjur.configuration.core_url, credentials)['host_factories'][id.identifier]['tokens']
+          RestClient::Resource.new(Conjur.configuration.core_url, Conjur.configuration.rest_client_options.merge(credentials || {}))['host_factories'][id.identifier]['tokens']
```
[the remark this answers is not in the document]
Comment (jtuttle) [reply lib/conjur/api/router/v4.rb:52]: Can we make a helper method for `Conjur.configuration.rest_client_options.merge(credentials || {})`? Looks like it's repeated a bunch.

File: lib/conjur/api/router/v5.rb [hunk 7]
```diff
@@ -27,19 +27,19 @@ module V5
         extend self
 
         def authn_login account, username, password
-          RestClient::Resource.new(Conjur.configuration.authn_url, user: username, password: password)[fully_escape account]['login']
+          RestClient::Resource.new(Conjur.configuration.authn_url, Conjur.configuration.rest_client_options.merge(user: username, password: password))[fully_escape account]['login']
```
Comment (jtuttle) [inline lib/conjur/api/router/v5.rb:30]: Please apply helper methods and breaking-up-lines liberally to this file. :)

File: spec/ssl_spec.rb [hunk 8]
```diff
@@ -16,25 +16,29 @@
 
   context 'with certificate added to the default OpenSSL cert store' do
     before do
-      store = OpenSSL::X509::Store.new
-      store.add_cert cert
-      stub_const 'OpenSSL::SSL::SSLContext::DEFAULT_CERT_STORE', store
+      cert_store.add_cert cert
```
Comment (jtuttle) [inline spec/ssl_spec.rb:19]: Add parens for method call `add_cert(cert)`

File: spec/configuration_spec.rb [hunk 9]
```diff
@@ -29,6 +29,28 @@
       configuration.account = "the-account"
       configuration.appliance_url = "https://conjur/api"
     }
+
+    it "rest_client_options defaults" do
+      expected = {
+        ssl_cert_store: OpenSSL::SSL::SSLContext::DEFAULT_CERT_STORE
+      }
+      expect(configuration.rest_client_options).to eq expected
```
Comment (jtuttle) [inline spec/configuration_spec.rb:37]: Add parens for method call: `eq(expected)`

File: spec/configuration_spec.rb [hunk 10]
```diff
@@ -29,6 +29,28 @@
...
+        ssl_ca_file: "ca_certificate.pem",
+        proxy: "http://proxy.example.com/"
+      }
+      configuration.rest_client_options = {
+        ssl_ca_file: "ca_certificate.pem",
+        proxy: "http://proxy.example.com/"
+      }
+
+      resource = Conjur::API.url_for(:authn_login, *["account", "username", "password"])
+      expect(resource.options).to include expected
```
Comment (jtuttle) [inline spec/configuration_spec.rb:51]: Add parens for method call: `include(expected)`

**Review by jtuttle — CHANGES_REQUESTED** [review]
Functionality seems fine, just some style requests.

File: lib/conjur/api/router/v4.rb [hunk 11]
```diff
@@ -29,35 +38,54 @@ def authn_authenticate_local username, account, expiration, cidr, &block
         def authn_rotate_api_key credentials, account, id
           verify_account(account)
           username = if id.kind == "user"
-            id.identifier
-          else
-            [ id.kind, id.identifier ].join('/')
-          end
-          RestClient::Resource.new(Conjur.configuration.authn_url, credentials)['users']["api_key?id=#{username}"]
+                       id.identifier
```
[the remark this answers is not in the document]
Comment (jtuttle) [reply lib/conjur/api/router/v4.rb:41]: I believe you can fix this slew of warnings by bumping the conditional onto the next line:

```
username =
  if id.kind ...
```

**Review by jtuttle — DISMISSED** [review]
Looks good, just a few easy (I think) codeclimate warnings to address.

**Review by jtuttle — APPROVED** [review]

**Pull request merged**